scipy/stats/_qmc.py

```diff
@@ -0,0 +1,1463 @@
+"""Quasi-Monte Carlo engines and helpers."""
+import copy
+import numbers
+from abc import ABC, abstractmethod
+import math
+import warnings
+
+import numpy as np
+
+import scipy.stats as stats
+from scipy.stats._sobol import (
+    initialize_v, _cscramble, _fill_p_cumulative, _draw, _fast_forward,
+    _categorize, initialize_direction_numbers, _MAXDIM, _MAXBIT
+)
+
+__all__ = ['scale', 'discrepancy', 'QMCEngine', 'Sobol', 'Halton',
+           'OrthogonalLatinHypercube', 'LatinHypercube',
+           'MultinomialQMC', 'MultivariateNormalQMC']
+
+
+# Based on scipy._lib._util.check_random_state
+def check_random_state(seed=None):
+    """Turn `seed` into a `numpy.random.Generator` instance.
+
+    Parameters
+    ----------
+    seed : {None, int, `numpy.random.Generator`,
+            `numpy.random.RandomState`}, optional
+        If `seed` is None the `numpy.random.Generator` singleton is used.
+        If `seed` is an int, a new ``Generator`` instance is used,
+        seeded with `seed`.
+        If `seed` is already a ``Generator`` or ``RandomState`` instance then
+        that instance is used.
+
+    Returns
+    -------
+    seed : {`numpy.random.Generator`, `numpy.random.RandomState`}
+        Random number generator.
+
+    """
+    if seed is None or isinstance(seed, (numbers.Integral, np.integer)):
+        if not hasattr(np.random, 'Generator'):
+            # This can be removed once numpy 1.16 is dropped
+            msg = ("NumPy 1.16 doesn't have Generator, use either "
+                   "NumPy >= 1.17 or `seed=np.random.RandomState(seed)`")
+            raise ValueError(msg)
+        return np.random.default_rng(seed)
+    elif isinstance(seed, np.random.RandomState):
+        return seed
+    elif isinstance(seed, np.random.Generator):
+        # The two checks can be merged once numpy 1.16 is dropped
+        return seed
+    else:
+        raise ValueError('%r cannot be used to seed a numpy.random.Generator'
+                         ' instance' % seed)
+
+
+def scale(sample, l_bounds, u_bounds, reverse=False):
+    r"""Sample scaling from unit hypercube to bounds range.
+
+    To convert a sample from :math:`[0, 1)` to :math:`[a, b), b>a`, the
+    following transformation is used:
+
+    .. math::
+
+        (b - a) \cdot \text{sample} + a
+
+    Parameters
+    ----------
+    sample : array_like (n, d)
+        Sample to scale.
+    l_bounds, u_bounds : array_like (d,)
+        Lower and upper bounds of transformed data. If `reverse` is True,
+        range of the original data to transform to the unit hypercube.
+    reverse : bool, optional
+        Reverse the transformation from `bounds` to the unit hypercube.
+        Default is False.
+
+    Returns
+    -------
+    sample : array_like (n, d)
+        Scaled sample.
+
+    Examples
+    --------
+    >>> from scipy.stats import qmc
```

Comment:
An example for which `n != d` might be easier to interpret. Explaining how the reverse transformation transforms back to the original `sample` and showing it would be a slam dunk.

Reply:
All done.